psbt script spend [3/3]: allow signing for asset script path spends through PSBTs

guggero opened the pull request:
~Depends on https://github.com/lightninglabs/taro/pull/243.~
Depends on https://github.com/lightninglabs/taro/pull/244.
~Depends on https://github.com/btcsuite/btcd/pull/1942.~

Fixes https://github.com/lightninglabs/taro/issues/75.

This PR contains the PSBT serialization logic for transporting asset related meta data in a BIP-174 based PSBT packet as well as all the required RPCs to allow creating and sending to a Taro address that locks assets to a script spend path.

jharveyb reviewed (commented):
Looking good! Surprisingly small changes to get script working. Only major suggested change is the FetchCommitment usage.

Will revisit to double-check the encode and Taproot logic.

Roasbeef commented on taropsbt/decode.go:
```diff
@@ -0,0 +1,377 @@
+	}
+
+	// Currently only one input is supported.
+	if len(packet.Inputs) != 1 {
```
Perhaps move this to a higher level validation layer (within the `AssetWallet` itself) so we don't need to go back and modify this later once we add additional inputs?

Roasbeef commented on taropsbt/decode.go:
```diff
@@ -0,0 +1,377 @@
+		key     []byte
+		decoder func([]byte) error
+	}{{
+		key:     PsbtKeyTypeInputTaroPrevID,
```
TLV within PSBT 👽

Def a fan of, as the OG PSBT serialization format is a bit wonky.

Roasbeef commented on tarodb/assets_store.go:
```diff
@@ -660,6 +660,34 @@ func constraintsToDbFilter(query *AssetQueryFilters) QueryAssetFilters {
+
+	if groupKey != nil {
+		key := groupKey.GroupPubKey
+		filter.KeyGroupFilter = key.SerializeCompressed()
```
isn't the asset ID enough to be able to uniquely pick out an asset?

guggero replied:
No, in case of a normal asset where I have multiple coins, we actually need the script key, which I added now. We don't necessarily need the group key for our specific query, but because this function creates a generic filter, I'm leaving that in for now.

Roasbeef commented on tarodb/assets_store.go:
```diff
@@ -1226,6 +1254,30 @@ func queryChainAssets(ctx context.Context, q ActiveAssetsStore,
+		return nil, fmt.Errorf("unable to query commitments: %w", err)
+	}
+
+	if len(commitments) != 1 {
```
Yeh see other comments re letting more than 1 go thru here, and then instead adding the temporary restriction at a higher level (within the wallet wallet itself).

guggero replied:
You're right, with the previous query it could've selected two coins of the same asset if they had the same value (or both a value greater than what we need). I replaced the amount with the script key in the query, which should only ever give one result (since the script key needs to be unique per anchor TX, I believe).

@Roasbeef at this level we are looking to get one specific asset input, independent of how many we allow on the virtual TX level.

Roasbeef commented on tarodb/assets_store.go:
```diff
@@ -1226,6 +1254,30 @@ func queryChainAssets(ctx context.Context, q ActiveAssetsStore,
 
+// FetchCommitment returns a specific commitment identified by the given asset
+// parameters. If no commitment is found, ErrNoCommitment is returned.
+func (a *AssetStore) FetchCommitment(ctx context.Context, id asset.ID,
```
Any reason this was added instead of using `SelectCommitment` which is more generic? Alternatively if a simpler API is needed elsewhere, then this can just call directly into that.

guggero replied:
The `SelectCommitment` didn't allow us to specify the asset we want precisely enough. There was always the chance that we'd get the wrong one if we had multiple asset UTXOs of the same value. I fixed that by adding the `FetchCommitment` method which now also takes the script key to make an asset UTXO identifiable.

Roasbeef commented on taroscript/send.go:
```diff
@@ -313,28 +312,24 @@ func SignVirtualTransaction(input *taropsbt.VInput, outputs []*taropsbt.VOutput,
-			return err
-		}
+	// TODO(guggero): Update this once we want to support more than one
+	// asset being spent in a single transaction.
```
See other comments re leaving the existing logic as it should work once we allow multiple inputs. We can also do simpler exercises of this by doing things like send two collectibles via an interactive flow in itests.

Roasbeef commented on taroscript/tx.go:
```diff
@@ -327,35 +329,116 @@ func InputScriptSpendSigHash(virtualTx *wire.MsgTx, input *asset.Asset,
+	// the control block to the witness, otherwise the verifier will reject
+	// the generated witness.
+	if spendDesc.SignMethod == input.TaprootScriptSpendSignMethod {
+		witness = append(witness, spendDesc.WitnessScript)
```
😈

Roasbeef commented on tarofreighter/chain_porter.go:
```diff
@@ -187,6 +187,25 @@ func (p *ChainPorter) RequestShipment(req *AssetParcel) (*PendingParcel, error)
+// RequestSignedDelivery requests a shipment of a virtual transaction that has
+// already been signed. This directly injects the deliverable into the state
+// machine at the state where it is anchored on chain.
+func (p *ChainPorter) RequestSignedDelivery(vPacket *taropsbt.VPacket,
```
Very cool that we're able to enable some external interaction with such a small added shim.

Roasbeef commented on tarofreighter/chain_porter.go:
```diff
@@ -229,28 +248,45 @@ func (p *ChainPorter) taroPorter() {
+				}
+
+			case req.pkg != nil:
+				sendPkg = req.pkg
```
Add log that this is the PSBT flow?

Roasbeef commented on tarofreighter/parcel.go:
```diff
@@ -79,6 +79,10 @@ type AssetParcel struct {
 
+	// pkg is the actual package that is being shipped. This is not set in
+	// case we request delivery to a destination address.
+	pkg *sendPackage
```
Perhaps use a distinct struct? Need to double check if `Dest` is still sent in this case, as otherwise might be some panics elsewhere since the field isn't set. 

We could also modify the actual channel use here to use a type param based on an interface that exposes the data needed via methods.

guggero replied:
Yeah, it wasn't really clear what fields would be accessed in what case. I refactored the parcels to make everything more generic and clear.

Roasbeef commented on rpcserver.go:
```diff
@@ -1040,6 +1064,133 @@ func (r *rpcServer) AddrReceives(ctx context.Context,
+	rpcsLog.Debugf("Selected commitment for anchor point %v, requesting "+
+		"delivery", inputCommitment.AnchorPoint)
+
+	resp, err := r.cfg.ChainPorter.RequestSignedDelivery(
```
So then this also assumes that the wallet as is will fully handle all the PSBT components? Could imagine another world where there's another external round to sign the Bitcoin portion (as that has a multi-sig/musig2 layer or w/e).

guggero replied:
Yes, definitely. The BTC level PSBT components will come in a later PR and give you full control over how the anchor TX is created, signed and published.

Roasbeef commented on tarorpc/assetwalletrpc/assetwallet.proto:
```diff
@@ -0,0 +1,140 @@
+        /*
+        Use an existing PSBT packet as the template for the funded PSBT.
+
+        TODO(guggero): Actually implement this. We can't use the "reserved"
```
Missing link here is parsing? Can def push that out to other PRs since this series itself is already somewhat hefty.

guggero replied:
Not just parsing, but a whole bunch of additional validations (e.g. only one input, first output set as change output, script keys set properly, coin selected properly and so on). So yeah, going to leave it for now.

guggero replied:
We'll start to support this to enable the interactive send case. But I decided to split that into a separate PR since more changes are necessary than initially expected.

Roasbeef commented on itest/psbt_test.go:
```diff
@@ -0,0 +1,180 @@
+
+	var (
+		alice = t.tarod
+		// aliceLnd = t.lndHarness.Alice
```
Stray comment?

Roasbeef commented on itest/psbt_test.go:
```diff
@@ -0,0 +1,180 @@
+		bobScriptKeyDesc.RawKeyBytes,
+	)
+	require.NoError(t.t, err)
+	bobInternalKeyDesc, err := bobLnd.WalletKitClient.DeriveNextKey(
```
As in re-expose the lnd API used to create keys in the first place?

guggero replied:
Yeah, I think it could be useful to have this call be threaded through the daemon so we don't need to know what key family to use in `lnd` when talking to `tarod`. Will add that to the PR that adds the BTC level PSBT stuff (for funding and signing the anchor transactions externally).

Roasbeef commented on itest/psbt_test.go:
```diff
@@ -0,0 +1,180 @@
+
+// testPsbtSend tests that we can properly send assets back and forth between
+// nodes with the use of PSBTs.
+func testPsbtSend(t *harnessTest) {
```
😎

Roasbeef reviewed (changes requested):
Excellent work! Gotta love when a few well crafted abstractions nicely fall into place. 

See comments in line, some carried over from the PR before it was split up. 

Few things popped out:
  * Ensuring that all the decoding + parsing is generic up until we restrict things to a single input deeper within the pipeline. 
  * Adding a similar coin selection mutex like we have in `lnd` over everything, which ensures that we don't have weird double spending issues. Alternatively, we can pipe everything through the porter and have it do single threaded coin selection + locking accounting. 
  * Update the test to also have a checksig in addition to the pre-image reveal so the vPSBT signing can be exercised there.

guggero commented on tarodb/assets_store.go:
```diff
@@ -660,6 +660,34 @@ func constraintsToDbFilter(query *AssetQueryFilters) QueryAssetFilters {
+		AssetIDFilter: id[:],
+		AnchorPoint:   anchorPointBytes,
+		MinAmt: sql.NullInt64{
+			Int64: int64(amt),
```
I removed the amount here and replaced it by the script key, as that should select the correct input asset with full precision (as compared to the amount).

dstadulis commented:
full value send test needs to be written.

guggero commented on taropsbt/interface.go:
```diff
@@ -0,0 +1,404 @@
+// We define a set of Taro specific global, input and output PSBT key types here
+// that correspond to the custom types defined in the VPacket below. We start at
+// 0x70 because that is sufficiently high to not conflict with any of the keys
+// specified in BIP174. Also, 7 is leet speak for "t" as in Taro.
```
Added two more sentences to the comment.

guggero commented on tarofreighter/chain_porter.go:
```diff
@@ -229,28 +248,45 @@ func (p *ChainPorter) taroPorter() {
+
+			// The request either has a destination address we want
+			// to send to, or a send package is already initialized.
+			switch {
```
I refactored the parcels into two distinct types to make this more clear.

guggero commented:
Thanks a lot for all the feedback, @Roasbeef and @jharveyb!

I addressed all of the comments and rebased on the latest version of part 2.

 
>     * Ensuring that all the decoding + parsing is generic up until we restrict things to a single input deeper within the pipeline.

Good idea, done!

 
>     * Adding a similar coin selection mutex like we have in `lnd` over everything, which ensures that we don't have weird double spending issues. Alternatively, we can pipe everything through the porter and have it do single threaded coin selection + locking accounting.

I agree. Though I think a mutex only adds any value if we can also mark a coin as being used/locked in the DB. I created an issue (https://github.com/lightninglabs/taro/issues/261) to track that.

 
>     * Update the test to also have a checksig in addition to the pre-image reveal so the vPSBT signing can be exercised there.

Yes, done.

I'm also going to split off the interactive send changes into a separate PR (I guess it will be [4/3] :sweat_smile: ), since the changes were a bit more involved than expected.

jharveyb commented on tarofreighter/parcel.go:
```diff
@@ -147,19 +94,138 @@ func (s SendState) String() string {
+		len(p.vPkt.Outputs))
+
+	// Initialize a package with the destination address.
+	return &sendPackage{
```
nit: inaccurate comment.

jharveyb reviewed (commented):
Looks good, only nits left IMO + changes in preceding PR.

guggero commented on itest/psbt_test.go:
```diff
@@ -0,0 +1,336 @@
+		},
+	)
+	require.NoError(t.t, err)
+	require.Contains(t.t, signedResp.SignedInputs, uint32(0))
```
Hmm, but since `SignVirtualPsbt` explicitly mentions virtual transactions in its name, I think it should be clear what the indexes refer to? Added a comment anyway.

guggero commented on tarogarden/mock.go:
```diff
@@ -36,7 +36,7 @@ func RandSeedlings(t testing.TB, numSeedlings int) map[string]*Seedling {
 			AssetName:      assetName,
 			Metadata:       test.RandBytes(32),
-			Amount:         uint64(rand.Int63()),
+			Amount:         uint64(rand.Int31()),
```
Cool, fixed it there as well.

guggero commented on taropsbt/decode.go:
```diff
@@ -0,0 +1,396 @@
+	var (
+		prevID            *asset.PrevID
+		anchorValue       uint64
+		anchorSigHashType uint64
```
Nice catch, fixed and test case updated!

guggero commented on taropsbt/encode.go:
```diff
@@ -0,0 +1,354 @@
+
+// encode encodes the current VInput struct into a PInput and a wire.TxIn.
+func (i *VInput) encode() (psbt.PInput, *wire.TxIn, error) {
+	txIn := &wire.TxIn{}
```
Good point, I removed it from the `encode()` method.

guggero commented on taroscript/send.go:
```diff
@@ -290,358 +176,564 @@ func IsValidInput(input *commitment.TaroCommitment,
+		// Any non change output shouldn't send to an un-spendable
+		// script, burning assets is done in a different way.
+		if !outputs[idx].IsChange &&
+			scriptPubKey.IsEqual(asset.NUMSPubKey) {
```
Yes, makes sense. Did it in part 2 though.

guggero commented:
Addressed all comments and added the input proof validation as well.

jharveyb commented on tarofreighter/wallet.go:
```diff
@@ -0,0 +1,717 @@
+		// Before we sign the transaction, we want to make sure the
+		// inclusion proof is valid and the asset is actually committed
+		// in the anchor transaction.
+		err := verifyInclusionProof(vPkt.Inputs[idx])
```
Probably want a TODO here to add a call to `ChainBridge` later on? To verify the anchor TX from the proof is in the chain.

guggero replied:
Great point! That also reminded me that we didn't check the anchor TX in the input vs. the one in the proof, which I added now. So only the "is the block in the chain" logic remains, for which I added a TODO.

jharveyb commented on tarofreighter/wallet.go:
```diff
@@ -0,0 +1,717 @@
+	}
+
+	inclusionProof := assetProof.InclusionProof
+	proofKey, _, err := inclusionProof.DeriveByAssetInclusion(
```
nit: have to check this err.

jharveyb reviewed (approved):
Looking good! Just have to remove `proof{Encoder,Decoder}`.

Roasbeef commented on taroscript/tx.go:
```diff
@@ -327,35 +329,123 @@ func InputScriptSpendSigHash(virtualTx *wire.MsgTx, input *asset.Asset,
+		}
+
+		spendDesc.SignMethod = input.TaprootScriptSpendSignMethod
+		spendDesc.TapTweak = vIn.TaprootMerkleRoot
```
Why is the tap tweak needed for a script spend from the tree? This is only needed for keyspend paths.

guggero replied:
Good point, removed.

Roasbeef commented on taroscript/tx.go:
```diff
@@ -327,35 +329,123 @@ func InputScriptSpendSigHash(virtualTx *wire.MsgTx, input *asset.Asset,
+	// One leaf hash and a merkle root means we're signing a specific
+	// script.
+	case len(vIn.TaprootMerkleRoot) == sha256.Size &&
+		len(derivation.LeafHashes) == 1:
```
Does this restrict to only a control block of size 1? So you only have a single sibling.

guggero replied:
No, we just only support signing for a single leaf at a time. The PSBT spec allows you to specify multiple leaf hashes and that would indicate to the signer that it should create a signature for each of them. But since we always know which path we're going to use, we haven't implemented this in `lnd` either. I updated the comment to make it more clear.

Roasbeef reviewed (approved):
LGTM 🌳

Just have one clarifying question (about the leaf hashes in the PSBT), and another re a field that doesn't need to be set for a script path spend (the tap tweak).
